packages/js/src/cache/notifications-cache.ts

```diff
@@ -0,0 +1,253 @@
+      if (!data) {
+        return;
+      }
+
+      const uniqueFilterKeys = new Set<string>();
+      this.#cache.keys().forEach((key) => {
+        data.forEach((notification) => {
+          let isNotificationFound = false;
+          if (remove) {
+            isNotificationFound = this.removeNotificationInCache(key, notification);
+          } else {
+            isNotificationFound = this.updateNotificationInCache(key, notification);
+          }
+
+          if (isNotificationFound) {
+            uniqueFilterKeys.add(getFilterKey(getFilter(key)));
+          }
+        });
+      });
+
+      uniqueFilterKeys.forEach((key) => {
+        const notificationsResponse = this.getAggregated(getFilter(key));
+        this.#emitter.emit('notifications.list.updated', {
+          data: notificationsResponse,
+        });
+      });
+    };
+
+  private has(args: ListNotificationsArgs): boolean {
+    return this.#cache.get(getCacheKey(args)) !== undefined;
+  }
+
+  private getAggregated(filter: NotificationFilter): ListNotificationsResponse {
+    const cacheKeys = this.#cache.keys().filter((key) => {
+      const parsedFilter = getFilter(key);
+
+      return isSameFilter(parsedFilter, filter);
+    });
+
+    return cacheKeys
+      .map((key) => this.#cache.get(key))
+      .reduce<ListNotificationsResponse>(
+        (acc, el) => {
+          if (!el) {
+            return acc;
+          }
+
+          return {
+            hasMore: el.hasMore,
+            filter: el.filter,
+            notifications: [...acc.notifications, ...el.notifications],
+          };
+        },
+        { hasMore: false, filter: {}, notifications: [] }
+      );
+  }
+
+  set(args: ListNotificationsArgs, data: ListNotificationsResponse): void {
+    this.#cache.set(getCacheKey(args), data);
+  }
+
+  getAll(args: ListNotificationsArgs): ListNotificationsResponse | undefined {
+    if (this.has(args)) {
+      return this.getAggregated({ tags: args.tags, read: args.read, archived: args.archived });
+    }
+
+    return;
+  }
+
+  /**
+   * Get unique notifications based on specified filter fields.
+   * The same tags can be applied to multiple filters which means that the same notification can be duplicated.
+   */
+  getUniqueNotifications({ tags, read }: Pick<ListNotificationsArgs, 'tags' | 'read'>): Array<Notification> {
+    const keys = this.#cache.keys();
+    const uniqueNotifications = new Map<string, Notification>();
+
+    keys.forEach((key) => {
+      const filter = getFilter(key);
+      if (areTagsEqual(tags, filter.tags)) {
+        const value = this.#cache.get(key);
+        if (!value) {
+          return;
+        }
+
+        value.notifications
+          .filter((el) => typeof read === 'undefined' || read === el.isRead)
+          .forEach((notification) => uniqueNotifications.set(notification.id, notification));
+      }
+    });
+
+    return Array.from(uniqueNotifications.values());
+  }
```

Comment:
This function is used for actions like: mark all read, mark all archived, mark read as archived. 
We need to collect a list of unique notifications for the specified filters and then optimistically update them. Later we send it with the pending event, so the cache can change and update it.